METRON-1911 Docker setup for testing bro plugin

ottobackwards opened the pull request:
Status>

`run_end_to_end.sh` is provided as an example of a testing script.  Specific or extended scripts can be created similar to this script to use the containers.
This script does the following:

1. Creates the Docker network
2. Runs the zookeeper container
3. Waits for zookeeper to be available
4. Runs the kafka container
5. Waits for kafka to be available
6. Creates the bro topic
7. Downloads sample PCAP data
8. Runs the bro container in the background

> Note that all parameters passed to this script are passed to the `docker_run_bro_container.sh` script

9. Builds the bro plugin
10. Configures the bro plugin
11. Runs bro against all the pcap data, one at a time
12. Executes a kafka client to read the data from bro for each pcap file
13. Stores the output kafka messages and the bro logs into the test_output directory

```bash
>tree Tue_Jan__8_21_54_10_EST_2019
Tue_Jan__8_21_54_10_EST_2019
├── exercise-traffic_pcap
│   ├── capture_loss.log
│   ├── conn.log
│   ├── dhcp.log
│   ├── dns.log
│   ├── files.log
│   ├── http.log
│   ├── kafka-output.log
│   ├── known_certs.log
│   ├── known_devices.log
│   ├── loaded_scripts.log
│   ├── notice.log
│   ├── packet_filter.log
│   ├── reporter.log
│   ├── smtp.log
│   ├── software.log
│   ├── ssl.log
│   ├── stats.log
│   ├── weird.log
│   └── x509.log
├── ftp_pcap
│   ├── capture_loss.log
│   ├── conn.log
│   ├── files.log
│   ├── ftp.log
│   ├── kafka-output.log
│   ├── loaded_scripts.log
│   ├── packet_filter.log
│   ├── reporter.log
│   ├── software.log
│   └── stats.log
```

14. Creates a results.csv for each pcap that has the line counts of the kafka and the bro output for each log
15. Prints all the results.csv to the screen

As we can see, the output is a folder named for the test run time, with a sub folder per pcap, containing all the bro logs and the kafka_output.log.


At this point the containers are up and running in the background.

Other scripts may then be used to do your testing, for example running:
```bash
./scripts/docker_execute_shell.sh
```

> NOTE: If the scripts are run repeatedly, and there is no change in bro or the librdkafka, the line `./run_end_to_end.sh` can be replaced by `./run_end_to_end.sh --skip-docker-build`, which uses the `--skip-docker-build` flag to not rebuild the bro container, saving the time of rebuilding bro and librdkafka.

> NOTE: After you are done, you must call the `finish_end_to_end.sh` script to cleanup.


##### `run_end_to_end.sh`
###### Parameters
```bash
--skip-docker-build            [OPTIONAL] Skip build of bro docker machine.
--data-path                    [OPTIONAL] The pcap data path. Default: ./data
```


## todo

 [x] start the bro container in the background
 [x] add host side scripts to docker exec 'foo'  where foo is a script in the bro container
 [x] execute the plugin build that way 
 [x] configure bro for kafka
 [x] run script to bro -r some pcaps
 [x] Cleanup and refactor
 [x] end to end 
 [x] output kafka messages to file
 [x] readme
 [x] output bro logs and kafka logs to same directory
 [x] break up kafka logs by bro log type
 [x] summarize line counts for logs in csv
 [x] show results in script

## Pull Request Checklist

Thank you for submitting a contribution to Apache Metron's Bro kafka writer plugin.

In order to streamline the review of the contribution we ask you follow these guidelines and ask you to double check the following:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? If not one needs to be created at [Metron Jira](https://issues.apache.org/jira/browse/METRON/?selectedTab=com.atlassian.jira.jira-projects-plugin:summary-panel).
- [x] Does your PR title start with METRON-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.
- [x] Has your PR been rebased against the latest commit within the target branch (typically master)?

### For code changes:
- [x] Have you included steps to reproduce the behavior or problem that is being changed or addressed?
- [x] Have you included steps or a guide to how the change may be verified and tested manually?
- [na] Have you ensured that the full suite of tests and checks have been executed via:
  ```
  bro-pkg test $GITHUB_USERNAME/metron-bro-plugin-kafka --version $BRANCH
  ```
- [-] Have you written or updated unit tests and or integration tests to verify your changes?
- [-] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [-] Have you verified the basic functionality of the build by building and running locally with Apache Metron's [Vagrant full-dev environment](https://github.com/apache/metron/tree/master/metron-deployment/development/centos6) or the equivalent?

ottobackwards commented:
I had a lot of trouble with compose, so I did it with plain run and a shared network.
We can look at compose later.

JonZeolla commented on docker/containers/bro-localbuild-container/Dockerfile:
```diff
@@ -0,0 +1,49 @@
+
+# install librdkafka
+RUN curl -L https://github.com/edenhill/librdkafka/archive/v0.9.4.tar.gz | tar xvz
+WORKDIR librdkafka-0.9.4/
```
This and above should be 0.11.5

ottobackwards replied:
all set

JonZeolla commented on docker/in_docker_scripts/build_bro.sh:
```diff
@@ -0,0 +1,31 @@
+#!/usr/bin/env bash
```
Shouldn't we rename the script, since this is building the plugin not bro.

ottobackwards replied:
all set

ottobackwards commented:
The idea, at least at this point is that when in the container you will have the 'built in scripts' that we have, and whatever scripts you have mapped into the container too.  the run_test.sh.template currently doesn't pass through the script dir parameter, I have to build it out at that top level.

So this will allow the user to write their own test scripts, maintained outside the code etc.

ottobackwards commented:
We will also have scripts that do docker exec etc to read from kafka etc.

JonZeolla commented on docker/in_docker_scripts/build_bro_plugin.sh:
```diff
@@ -0,0 +1,39 @@
+
+cd /root || exit 1
+echo "================================" >> "${RUN_LOG_PATH}" 2>&1
+bro-pkg install code --force | tee "${RUN_LOG_PATH}"
```
Just for clarity, would you mind changing this to `./code`?

JonZeolla replied:
Should we also capture the rc and exit if != 0?

JonZeolla commented on docker/in_docker_scripts/wait-for-it.sh:
```diff
@@ -0,0 +1,161 @@
+#!/usr/bin/env bash
+#   Use this script to test if a given TCP host/port are available
+
+cmdname=$(basename $0)
```
Suggest `cmdname="$( cd "$( dirname "${BASH_SOURCE[0]}" )" >/dev/null && pwd )"` as a general good practice.  Probably no specific need if this works, since it's a somewhat closed environment.

JonZeolla replied:
Agreed, since it isn't ours.

JonZeolla commented on docker/scripts/consume_bro_kafka.sh:
```diff
@@ -0,0 +1,22 @@
+
+shopt -s nocasematch
+docker run --rm --network bro-network ches/kafka \
+kafka-console-consumer.sh --topic bro --from-beginning --bootstrap-server kafka:9092
```
Probably indent to make it more clear it's a multiline command.

JonZeolla commented on docker/scripts/create_docker_network.sh:
```diff
@@ -0,0 +1,73 @@
+
+docker network create "${NETWORK_NAME}"
+rc=$?; if [[ ${rc} != 0 ]]; then
+exit ${rc};
```
No need for `;`, also should indent

JonZeolla commented on docker/scripts/destroy_docker_network.sh:
```diff
@@ -0,0 +1,71 @@
+
+docker network rm "${NETWORK_NAME}"
+rc=$?; if [[ ${rc} != 0 ]]; then
+exit ${rc};
```
No need for `;`, also should indent

JonZeolla commented on docker/scripts/download_sample_pcaps.sh:
```diff
@@ -0,0 +1,101 @@
+done
+
+if [[ ! -f ${DATA_PATH}/example-traffic/exercise-traffic.pcap ]]; then
+  wget https://www.bro.org/static/traces/exercise-traffic.pcap -O ${DATA_PATH}/example-traffic/exercise-traffic.pcap
```
Repoint to our repo.

JonZeolla replied:
github.com/apache/metron-bro-plugin-kafka, since you've added the pcaps under docker/data

JonZeolla commented on docker/scripts/download_sample_pcaps.sh:
```diff
@@ -0,0 +1,101 @@
+  wget https://github.com/EmpowerSecurityAcademy/wireshark/blob/master/radius_localhost.pcapng?raw=true -O ${DATA_PATH}/radius/radius_localhost.pcapng
+fi
+if [[ ! -f ${DATA_PATH}/rfb/rfb.pcap ]]; then
+  wget https://github.com/kholia/my-pcaps/blob/master/VNC/07-vnc-openwall-3.7.pcap?raw=true -O ${DATA_PATH}/rfb/rfb.pcap
```
Repoint to our repo.

JonZeolla commented on docker/scripts/download_sample_pcaps.sh:
```diff
@@ -0,0 +1,101 @@
+  wget https://github.com/markofu/pcaps/blob/master/PracticalPacketAnalysis/ppa-capture-files/ftp.pcap?raw=true -O ${DATA_PATH}/ftp/ftp.pcap
+fi
+if [[ ! -f ${DATA_PATH}/radius/radius_localhost.pcapng ]]; then
+  wget https://github.com/EmpowerSecurityAcademy/wireshark/blob/master/radius_localhost.pcapng?raw=true -O ${DATA_PATH}/radius/radius_localhost.pcapng
```
Repoint to our repo.

JonZeolla replied:
We should include this as a pcap, not pcapng.  Run `editcap -F libpcap radius_localhost.pcapng radius_localhost.pcap` on the pcap, then replace it in the repo.

ottobackwards replied:
as above I'm not sure which repo you mean.

JonZeolla commented on docker/scripts/download_sample_pcaps.sh:
```diff
@@ -0,0 +1,101 @@
+  wget https://www.bro.org/static/traces/ssh.pcap -O ${DATA_PATH}/ssh/ssh.pcap
+fi
+if [[ ! -f ${DATA_PATH}/ftp/ftp.pcap ]]; then
+  wget https://github.com/markofu/pcaps/blob/master/PracticalPacketAnalysis/ppa-capture-files/ftp.pcap?raw=true -O ${DATA_PATH}/ftp/ftp.pcap
```
Repoint to our repo.

JonZeolla commented on docker/scripts/download_sample_pcaps.sh:
```diff
@@ -0,0 +1,101 @@
+  wget http://downloads.digitalcorpora.org/corpora/network-packet-dumps/2008-nitroba/nitroba.pcap -O ${DATA_PATH}/nitroba/nitroba.pcap
+fi
+if [[ ! -f ${DATA_PATH}/ssh/ssh.pcap ]]; then
+  wget https://www.bro.org/static/traces/ssh.pcap -O ${DATA_PATH}/ssh/ssh.pcap
```
Repoint to our repo.

JonZeolla commented on docker/scripts/download_sample_pcaps.sh:
```diff
@@ -0,0 +1,101 @@
+  wget https://www.bro.org/static/traces/exercise-traffic.pcap -O ${DATA_PATH}/example-traffic/exercise-traffic.pcap
+fi
+if [[ ! -f ${DATA_PATH}/nitroba/nitroba.pcap ]]; then
+  wget http://downloads.digitalcorpora.org/corpora/network-packet-dumps/2008-nitroba/nitroba.pcap -O ${DATA_PATH}/nitroba/nitroba.pcap
```
Repoint to our repo.

ottobackwards commented on docker/in_docker_scripts/process_data_dir.sh:
```diff
@@ -0,0 +1,32 @@
+  exit 1
+fi
+
+ls /root/data
```
no debugging

JonZeolla replied:
So shouldn't we remove it?

ottobackwards replied:
sure

ottobackwards commented on docker/in_docker_scripts/wait-for-it.sh:
```diff
@@ -0,0 +1,161 @@
+#!/usr/bin/env bash
+#   Use this script to test if a given TCP host/port are available
```
it is from our metron-contrib/docker, he may have gotten it from there.  I'll have to check the L&N

ottobackwards replied:
all set

ottobackwards commented on docker/run_test.sh.template:
```diff
@@ -0,0 +1,199 @@
+#!/usr/bin/env bash
```
so the idea is that this is a sample, that they would modify, and would not get committed over and over again

ottobackwards replied:
i changed it

JonZeolla commented on docker/scripts/process_data_docker.sh:
```diff
@@ -0,0 +1,29 @@
+
+shopt -s nocasematch
+
+echo "exectuting proces_data_dir.sh in the bro docker container"
```
"executing process_data_dir.sh"

ottobackwards replied:
fixed

JonZeolla commented:
Is there a reason the scripts in `scripts/` have inconsistent permissions?  Should `example_script.sh` have execute, or maybe even just call it `run_test.sh` off the bat?

ottobackwards commented:
@JonZeolla , I was going to ask you about that, what do you think the best practice is for these permissions?

JonZeolla commented on docker/example_script.sh:
```diff
@@ -0,0 +1,177 @@
+if [[ "$SKIP_REBUILD_BRO" = false ]] ; then
+  bash "${SCRIPT_DIR}"/build_container.sh \
+    --container-directory="${CONTAINER_DIR}" \
+    --container-name=bro-docker-container:latest
```
We should include metron in the container tag.

ottobackwards replied:
done

ottobackwards commented on docker/example_script.sh:
```diff
@@ -0,0 +1,177 @@
+function shutdown {
+
+  if [[ "$RAN_BRO_CONTAINER" = true ]]; then
+    "${SCRIPT_DIR}"/stop_container.sh --container-name=bro
```
done

JonZeolla replied:
I don't think this was done.

ottobackwards replied:
the container name is not variable here, even though there is a var defined it isn't passed

ottobackwards commented on docker/scripts/cleanup_containers.sh:
```diff
@@ -0,0 +1,30 @@
+
+SCRIPT_DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" >/dev/null && pwd )"
+
+"${SCRIPT_DIR}"/stop_container.sh --container-name=bro
```
done

ottobackwards commented on docker/example_script.sh:
```diff
@@ -0,0 +1,177 @@
+#run the bro container
+#and optionally the passed script _IN_ the container
+bash "${SCRIPT_DIR}"/run_bro_container.sh --container-path="${CONTAINER_DIR}" \
+  --container-name=bro-docker-container:latest \
```
done

ottobackwards commented on docker/in_docker_scripts/build_bro_plugin.sh:
```diff
@@ -0,0 +1,39 @@
+echo "redef Kafka::logs_to_exclude = set(Conn::LOG, DHCP::LOG);" >> /usr/local/bro/share/bro/site/local.bro
+echo "redef Known::cert_tracking = ALL_HOSTS;" >> /usr/local/bro/share/bro/site/local.bro
+echo "redef Software::asset_tracking = ALL_HOSTS;" >> /usr/local/bro/share/bro/site/local.bro
+sed -i '86 a @load policy/protocols/dhcp/known-devices-and-hostnames.bro' /usr/local/bro/share/bro/site/local.bro
```
done

ottobackwards commented:
current send to end cmd is:

```bash
./scripts/download_sample_pcaps.sh --data-path=A_FULL_PATH &&  ./example_script.sh --skip-docker-build --leave-running --data-path=A_FULL_PATH && ./scripts/docker_execute_process_data_dir.sh &&  ./scripts/docker_run_consume_bro_kafka.sh
```

ottobackwards commented:
@nickwallen @JonZeolla, with the readme done, this is now ready for review

ottobackwards commented:
@nickwallen @JonZeolla, with the readme done, this is now ready for review

nickwallen commented:
Thanks @ottobackwards .  I'll give it a run through.

nickwallen commented:
I ran this up, but seemed to get a couple errors.
```
./scripts/docker_run_bro_container.sh: line 152: DOCKER_CMD_BASE: command not found
```
```
./scripts/docker_run_bro_container.sh: line 162: metron-bro-docker-container:latest: command not found
```

This is the command that I ran along with what I saw when the errors popped up.
```
$ ./scripts/download_sample_pcaps.sh --data-path=~/tmp/data  &&  ./example_script.sh  --leave-running --data-path=/Users/me/tmp/pcap_data && ./scripts/docker_execute_process_data_dir.sh &&  ./scripts/docker_run_consume_bro_kafka.sh

...

make[1]: Leaving directory `/root/bro-2.5.5/librdkafka-0.11.5/src-cpp'
Removing intermediate container c4b01dee8e27
 ---> 4301bf29af13
Step 23/23 : WORKDIR /root
 ---> Running in 3fac0790cfdc
Removing intermediate container 3fac0790cfdc
 ---> 02b3e630af02
Successfully built 02b3e630af02
Successfully tagged metron-bro-docker-container:latest
Running docker_run_bro_container with
CONTAINER_NAME = bro
NETWORK_NAME = bro-network
SCRIPT_PATH =
LOG_PATH = /Users/nallen/tmp/metron-bro-plugin-kafka-pr21/docker/logs
DATA_PATH = /Users/me/tmp/pcap_data
DOCKER_PARAMETERS =
===================================================
Log will be found on host at /Users/nallen/tmp/metron-bro-plugin-kafka-pr21/docker/logs/bro-test-Fri_Dec__7_16:51:06_EST_2018.log
./scripts/docker_run_bro_container.sh: line 143: DOCKER_CMD_BASE: command not found
./scripts/docker_run_bro_container.sh: line 144: DOCKER_CMD_BASE: command not found
./scripts/docker_run_bro_container.sh: line 145: DOCKER_CMD_BASE: command not found
./scripts/docker_run_bro_container.sh: line 146: DOCKER_CMD_BASE: command not found
./scripts/docker_run_bro_container.sh: line 147: DOCKER_CMD_BASE: command not found
./scripts/docker_run_bro_container.sh: line 152: DOCKER_CMD_BASE: command not found
===============Running Docker===============

eval command is:
bro bash

============================================

./scripts/docker_run_bro_container.sh: line 162: metron-bro-docker-container:latest: command not found
Running stop_container with
CONTAINER_NAME= kafka
===================================================
kafka
kafka
Running stop_container with
CONTAINER_NAME= zookeeper
===================================================
zookeeper
zookeeper
Running destroy_docker_network with
NETWORK_NAME = bro-network
===================================================
bro-network
```

And I see no containers running.
```
$ docker ps
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS              PORTS               NAMES
```

Docker Engine 18.09.0

ottobackwards commented:
ok, this is fixed.  The bashSupport formatter in IntelliJ break the array assignment.  Sorry, i didn't redeploy after a format and  document run.

Also, I don't think your pcap directories match.   And- I have found I need to pass full paths for both of those scripts/parameters.

When I ran your line I ended up downloading to  a folder '~' in my current directory.

Sorry for the trouble.  If you could suggest a good bash formatter that would be super.

nickwallen commented on docker/in_docker_scripts/build_bro_plugin.sh:
```diff
@@ -0,0 +1,48 @@
+cd /root || exit 1
+
+echo "================================" >>"${RUN_LOG_PATH}" 2>&1
+bro-pkg install code --force | tee "${RUN_LOG_PATH}"
```
Why not install the plugin in the Dockerfile?  That way an image with the plugin installed and ready to go can be cached.

ottobackwards replied:
Because one of the uses of this is to regression test changes in the plugin itself.  Therefore you may _not_ rebuild the bro image, but want to rebuild the plugin

nickwallen commented on docker/scripts/download_sample_pcaps.sh:
```diff
@@ -0,0 +1,105 @@
+done
+
+if [[ ! -f "${DATA_PATH}"/example-traffic/exercise-traffic.pcap ]]; then
+  wget https://www.bro.org/static/traces/exercise-traffic.pcap -O "${DATA_PATH}"/example-traffic/exercise-traffic.pcap
```
Why not include this in the Dockerfile?  That way it can be cached in the image.

ottobackwards replied:
I had thought of the pcaps as being external to the images, that you may have different sets of pcaps that you want to test with, so leaving that configurable would be a good thing

nickwallen replied:
I don't see the data sets that I want to test with changing all that much. What I do see, is a developer making changes to the plugin and wanting to spin those changes up quickly.  Having the pcaps cached in the image is going to help make that iteration much quicker.

nickwallen replied:
This is not a big deal though.  What you have should work.  We could probably use a data container, but no sense in optimizing the heck out of this right now.

ottobackwards commented:
how are we feeling on this?

JonZeolla commented:
On my backlog to continue reviewing.  May not be able to until tomorrow but I'll get all the containers built and such today to prep

nickwallen commented:
I need to find time to try and get this working.  Am still running into the problems I described above.

ottobackwards commented:
@nickwallen do you have any suggestions for default directories for that case then?  I wanted to leave it configurable, the idea being that you would have *your* version of the script for your environment

ottobackwards commented:
it, would defaulting ~/tmp/metron-bro-kafka-run/data|logs or something?

nickwallen commented:
> @ottobackwards said: do you have any suggestions for default directories for that case then? I wanted to leave it configurable, the idea being that you would have your version of the script for your environment

The configurability is nice.  All the start script would do is just run all those options.

Could we just use a data directory in the source tree;  `docker/data`?  And just .gitignore it?  Makes the least assumptions about a user's environment.

ottobackwards commented:
Ok, I'll default to that and make it configurable.  I can imagine having multiple pcap sets and not wanting to copy them around etc.

ottobackwards commented:
@nickwallen @JonZeolla please see the latest commits, updated PR description and README.
I have refactored it such that there is a working end to end test script that outputs kafka messages to log, and a finish script to cleanup.

JonZeolla commented:
Don't forget that in order to install this in a bro cluster the repo needs cloned.  I'm a bit hesitant to suggest that in every case those pcaps should be a part of that.  I prefer the dynamic download w/existence check for that reason.

JonZeolla commented:
Regarding Q2, when I ran this up I saw it segfault as expected after each pcap was done processing.

nickwallen commented:
> @JonZeolla: Regarding Q2, when I ran this up I saw it segfault as expected after each pcap was done processing.

You just had to watch the output as it progressed to see this?  It did not stop processing mid-stream because of the seg fault?

JonZeolla commented:
For Q3 I was thinking a quick win would be making sure the number of messages pulled off of kafka lines up with the number of log lines written to disk by bro in .log files (after removing the standard bro log file header, or we change the bro config to write JSON to disk and just do a 1:1 comparison).  Of course we could take it further and look at content as well.  I don't think that necessarily needs to be a part of this PR although it could be.

nickwallen commented:
> @JonZeolla: For Q3 I was thinking a quick win would  would be making sure the number of messages pulled off of kafka ...

Yeah, I think we could do a couple things to make this more "test-like" so we can run it as part of a CI build.  I agree with you that what @ottobackwards has here should go in as-is, rather than add scope.

nickwallen commented:
I see it now, just a large amount of output to pick through.
```
processing /root/data/rfb/rfb.pcap
1328634261.353218 stats/Log::WRITER_KAFKAWRITER: Debug is turned off.
1328634274.303408 rfb/Log::WRITER_KAFKAWRITER: Debug is turned off.
1328634276.909530 capture_loss/Log::WRITER_KAFKAWRITER: Debug is turned off.
find: 'bro' terminated by signal 11
```

nickwallen commented:
(Q4). So say I am working on enhancing the plugin and I want to test the changes that I've made using this.  What commands do I need to run to do that?  

I assume there is a more efficient way than running end-to-end each time.  Just building Bro itself takes a significant amount of time.

nickwallen commented:
+1 Looks good to me.  Thanks!

ottobackwards commented:
@JonZeolla are you all set?

JonZeolla commented:
Sorry, not yet.  Can look tomorrow

JonZeolla commented:
Have you considered adding something like `set -o errexit -o errtrace -o pipefail -o nounset` in the scripts?

JonZeolla commented on docker/in_docker_scripts/configure_bro_plugin.sh:
```diff
@@ -0,0 +1,41 @@
+} >> /usr/local/bro/share/bro/site/local.bro
+
+
+sed -i '86 a @load policy/protocols/dhcp/known-devices-and-hostnames.bro' /usr/local/bro/share/bro/site/local.bro
```
This will error with bro 2.6 and newer, and generally isn't a very clean appraoch.  In 2.6, the DHCP framework was reworked and `known-devices-and-hostnames.bro` no longer exists (https://github.com/zeek/zeek/commit/e76b56ce530aff67422ab24c66c4de2ed8b467d7#diff-cdafd4a1100bf5b58e63223c5d25a906).  This command was initially made as a hack to test Metron ES support for bro 2.5.5 logs.  I suggest we figure out what the right approach here is - don't have anything off the top of my head right now.

nickwallen replied:
Since this does work for 2.5.x, I'd suggest we just get this PR merged.  We can tackle support for 2.6 in a separate PR.  I'd hate for this PR to continue to drag on.  But good eye @JonZeolla .

ottobackwards replied:
this isn't only in the new script.  we should tackle that when we do the rest of the 2.6 stuff

JonZeolla replied:
I agree, we're getting very close.  My review is almost done.  Just want to make sure this one doesn't get missed - added to the existing [METRON-1914](https://issues.apache.org/jira/browse/METRON-1914?filter=-3)

JonZeolla commented on docker/scripts/download_sample_pcaps.sh:
```diff
@@ -0,0 +1,106 @@
+  wget https://github.com/kholia/my-pcaps/blob/master/VNC/07-vnc-openwall-3.7.pcap?raw=true -O "${DATA_PATH}"/rfb/rfb.pcap
+fi
+
+rc=$?; if [[ ${rc} != 0 ]]; then
```
What is this supposed to be checking?

JonZeolla commented:
One last thing - can we run `bro -r` from pcap-unique directories, so it will leave `.log` artifacts per-pcap instead of overwriting any prior files in `/root` as it currently is doing?

ottobackwards commented:
ok, i see what you are saying.  I never checked that output:

```bash
scripts/docker_execute_shell.sh
Running build_bro_plugin_docker with
CONTAINER_NAME = bro
===================================================
[root@def60f96cb67 ~]# ls
anaconda-ks.cfg   capture_loss.log  data      dpd.log    http.log           loaded_scripts.log  packet_filter.log  rfb.log   software.log  stats.log
bro-2.5.5         code              dhcp.log  files.log  known_certs.log    logs                radius.log         sip.log   ssh.log       weird.log
built_in_scripts  conn.log          dns.log   ftp.log    known_devices.log  notice.log          reporter.log       smtp.log  ssl.log       x509.log
[root@def60f96cb67 ~]# pwd
/root
[root@def60f96cb67 ~]#
```

ottobackwards commented:
How about if we had a bro_logs directory?

ottobackwards commented:
or do you need $BRO_LOG_DIR/PCAP_NAME/*.log?

JonZeolla commented:
The problem is that every `bro -r` will produce some overlapping logs like conn.log.  The last run will win and overwrite any prior information.

JonZeolla commented:
Each pcap, yes

ottobackwards commented:
ok, that makes sense.  Sorry I missed that, I didn't think about the bro log at all, only the kafka

ottobackwards commented:
This feels like it is circling the 'right' thing.  How to output the logs for each run in a way that you can work with them.   Since I expect users to do some shell work around this, it is ok.  But we are prob. going to refactor this after usage.  which is OK

ottobackwards commented:
I am going to do a refactor to output a directory with all the outputs for the run paired ( bro-log-xxx -> kafka output xxxx ).

ottobackwards commented:
```bash
ree Tue_Jan__8_21_54_10_EST_2019
Tue_Jan__8_21_54_10_EST_2019
├── exercise-traffic_pcap
│   ├── capture_loss.log
│   ├── conn.log
│   ├── dhcp.log
│   ├── dns.log
│   ├── files.log
│   ├── http.log
│   ├── kafka-output.log
│   ├── known_certs.log
│   ├── known_devices.log
│   ├── loaded_scripts.log
│   ├── notice.log
│   ├── packet_filter.log
│   ├── reporter.log
│   ├── smtp.log
│   ├── software.log
│   ├── ssl.log
│   ├── stats.log
│   ├── weird.log
│   └── x509.log
├── ftp_pcap
│   ├── capture_loss.log
│   ├── conn.log
│   ├── files.log
│   ├── ftp.log
│   ├── kafka-output.log
│   ├── loaded_scripts.log
│   ├── packet_filter.log
│   ├── reporter.log
│   ├── software.log
│   └── stats.log
```

This is really nice, I think it is what you are looking for @JonZeolla

ottobackwards commented:
I don't know about splitting the kafka log per log type....

JonZeolla commented:
Yup that's exactly it.  Thanks.  Kafka per log is nice to have but not necessary, I think we can work with what we've got/handle as a follow on.  But I wouldn't say no.  :)

ottobackwards commented:
So with the recent changes:

1.  everything get's output together in the test_output directory ( see the readme )
2. there is the main kafka-output.log as well as a $LOGNAME.kafka.log for each log
3. there is a results.csv per pcap as well that has the line counts for kafka and bro per log
4 the script prints out the results:

```bash
--> /Users/ottofowler/src/apache/forks/metron-bro-plugin-kafka/docker/test_output/Wed_Jan__9_10_28_09_EST_2019/ssh_pcap/results.csv
LOG           BRO_COUNT  KAFKA_COUNT
capture_loss         1         58
notice               1         53
software             4        987
ssh                 40         40
stats                2        162
========================================================

--> /Users/ottofowler/src/apache/forks/metron-bro-plugin-kafka/docker/test_output/Wed_Jan__9_10_28_09_EST_2019/radius_localhost_pcapng/results.csv
LOG           BRO_COUNT  KAFKA_COUNT
capture_loss         1         56
radius               7          7
stats                2        159
========================================================
```

So with the csv etc I think everything is setup now to adapt this for more specific tests etc.

> NOTE: The numbers for the the non-protocol logs don't match up well.  I think there may be output in the log that doesn't go into kafka.  I already try to escape lines that start with `#`, but there may be more

Please have a look @nickwallen @JonZeolla

JonZeolla commented:
This is slick, nice work.  Give me some time to review

ottobackwards commented:
I'm going to need some help on the count differences.  A lot of them are the same, maybe that can be a follow up?

ottobackwards commented on docker/scripts/docker_run_bro_container.sh:
```diff
@@ -18,6 +18,9 @@
 #
 
 shopt -s nocasematch
+set -e # errexit
```
the nounset was making the array parameters handling break, so I took it out

ottobackwards commented on docker/scripts/split_kakfa_output_by_log.sh:
```diff
@@ -0,0 +1,103 @@
+#  limitations under the License.
+#
+
+shopt -s nocasematch
```
done

ottobackwards commented:
the logs/ *  that I tried to setup to have docker log stuff.... just doesn't work.
I'm going to remove it

JonZeolla commented:
Has this stabilized enough for a final review?

ottobackwards commented:
yes, adding jq was the last thing.

JonZeolla commented on docker/scripts/docker_run_bro_container.sh:
```diff
@@ -0,0 +1,168 @@
+declare -a DOCKER_CMD_BASE
+DOCKER_CMD="bash"
+DOCKER_CMD_BASE[0]="docker run -d -t --name ${CONTAINER_NAME} --network ${NETWORK_NAME} "
+DOCKER_CMD_BASE[2]="-v \"${OUR_SCRIPTS_PATH}:/root/built_in_scripts\" "
```
Is there a reason we're skipping [1] here?  I believe it used to be logs, which got removed recently.

JonZeolla commented:
Please align the PR title and the JIRA

JonZeolla commented on docker/README.md:
````diff
@@ -0,0 +1,283 @@
+└── wait_for_zk.sh
+```
+
+- `build_bro_plugin.sh`: Runs `bro-package` to build and install the plugin.
````
`s/bro-package/bro-pkg/`

JonZeolla commented on docker/README.md:
````diff
@@ -0,0 +1,283 @@
+
+  ###### Parameters
+  ```bash
+  --data-path                    [REQURIED] The pcap data path
````
`s/REQURIED/REQUIRED/`

JonZeolla commented on docker/scripts/download_sample_pcaps.sh:
```diff
@@ -0,0 +1,112 @@
+function help {
+  echo " "
+  echo "usage: ${0}"
+  echo "    --data-path                    [REQURIED] The pcap data path"
```
`s/REQURIED/REQUIRED/`

JonZeolla commented on docker/README.md:
````diff
@@ -0,0 +1,283 @@
+- `print_results.sh` : Prints the `results.csv` for all the pcaps processed in the given directory to console
+  ###### Parameters
+  ```bash
+  --test-directory               [REQUIRED] The directory for the tests"
````
`'s/"//'`

JonZeolla commented on docker/README.md:
````diff
@@ -0,0 +1,283 @@
+- `split_kafka_output_by_log.sh` : For a pcap result directory, will create a LOG.kafka.log for each LOG.log's entry in the kafka-output.log
+  ###### Parameters
+  ```bash
+  --log-directory                [REQUIRED] The directory with the logs"
````
`'s/"//'`

JonZeolla commented on docker/run_end_to_end.sh:
```diff
@@ -0,0 +1,197 @@
+  bash "${SCRIPT_DIR}"/docker_execute_process_data_file.sh --pcap-file-name="${BASE_FILE_NAME}" --output-directory-name="${DOCKER_DIRECTORY_NAME}"
+
+  rc=$?; if [[ ${rc} != 0 ]]; then
+    echo "ERROR> FAILED TO PROCESS ${file} DATA.  CHECK LOGS  ${rc}, please run the finish_end_to_end.sh when you are done."
```
Is this meant to say `CHECK LOGS  ${rc}`?  Does feel right, did you mean to point to `${TEST_OUTPUT_PATH}/${DOCKER_DIRECTORY_NAME}`?

ottobackwards replied:
Cleaned it up.  Yes, I meant that.

JonZeolla commented on docker/in_docker_scripts/build_bro_plugin.sh:
```diff
@@ -0,0 +1,43 @@
+shopt -s nocasematch
+
+#
+# Runs bro-package to build and install the plugin
```
`s/bro-package/bro-pkg/`

JonZeolla commented:
It appears that the kafka logs are somehow getting accumulated across each pcap run.  Here is an example using `capture_loss`:

```
> for x in `ls`; do echo $x;cat $x/results.csv; done | grep capture_loss
capture_loss,      41,      41
capture_loss,       1,      42
capture_loss,      13,      55
capture_loss,       1,      56
capture_loss,       1,      57
capture_loss,       1,      58
```

JonZeolla commented on docker/scripts/docker_run_consume_bro_kafka.sh:
```diff
@@ -0,0 +1,75 @@
+done
+
+docker run --rm --network "${NETWORK_NAME}" ches/kafka \
+  kafka-console-consumer.sh --topic bro --from-beginning --bootstrap-server kafka:9092 --timeout-ms 1000
```
Is it possible that the duplication issue we're seeing is due to `--from-beginning` here?  Would need to look and see if this script is being run for each pcap or not.

ottobackwards replied:
fixed

JonZeolla commented on docker/run_end_to_end.sh:
```diff
@@ -0,0 +1,197 @@
+  fi
+
+  KAFKA_OUTPUT_FILE="${TEST_OUTPUT_PATH}/${DOCKER_DIRECTORY_NAME}/kafka-output.log"
+  bash "${SCRIPT_DIR}"/docker_run_consume_bro_kafka.sh | "${ROOT_DIR}"/remove_timeout_message.sh | tee "${KAFKA_OUTPUT_FILE}"
```
Yup, this is being run for each pcap, and it goes from the beginning, which is causing the duplciation.

JonZeolla commented:
+1 This looks great, I reran it and the numbers are much more similar (although not perfect).  This likely indicates an issue with master and not in this PR.  Specifically:

```--> /Users/jzeolla/tmp/metron-bro-plugin-kafka-pr21/docker/test_output/Tue_Feb__5_12_49_14_EST_2019/nitroba_pcap/results.csv
LOG           BRO_COUNT  KAFKA_COUNT
http              4511       4545
x509                46         96
========================================================

--> /Users/jzeolla/tmp/metron-bro-plugin-kafka-pr21/docker/test_output/Tue_Feb__5_12_49_14_EST_2019/exercise-traffic_pcap/results.csv
LOG            BRO_COUNT  KAFKA_COUNT
http               2514       2517
x509                 56        108
```

@nickwallen does your +1 still hold from [here](https://github.com/apache/metron-bro-plugin-kafka/pull/21#issuecomment-452029698)?

Great job @ottobackwards

JonZeolla commented:
Noteworthy:  I reran this from scratch and got the exact same output - same pcaps caused the same counts in the BRO_COUNT and KAFKA_COUNT columns.

JonZeolla commented:
Here are some tickets for follow-on work:
- [METRON-1991](https://issues.apache.org/jira/browse/METRON-1991)
- [METRON-1990](https://issues.apache.org/jira/browse/METRON-1990)
- [METRON-1914](https://issues.apache.org/jira/browse/METRON-1914)

nickwallen commented:
I am a +0 on this.  I just haven't had time to review since the first time I did.  And since that time many changes have been made.

I think we should merge it though, based on @JonZeolla 's +1.  Thanks for all the hard work @ottobackwards !

nickwallen commented:
And most critically, I was waiting on this to land so that we could start looking at #20 and [METRON-1910](https://issues.apache.org/jira/browse/METRON-1910).

ottobackwards commented:
thanks!
